### ISSUE TITLE
IoT Edge module stops receiving data

### ISSUE THREAD

**Issue by draganrmijailovic** [Azure/iotedge#5065]
## Expected Behavior
The edgeAgent and the whole runtime runs reliably and stable, without interruptions.

## Current Behavior
We have created an IoT Edge module for receiving data from the devices. Our module sends data as messages to IoT Hub and via our event hub triggered Azure function we store messages in the database. I created about fifteen days ago: https://github.com/Azure/iotedge/issues/4992. The issue was closed but the problem, however, has not been solved.  But, now I have new information:

**The IoT edge module stops receiving messages after a certain time. I improved the module for receiving messages and created two threads - one thread receives messages and stores them in the BlockingCollection. Another thread reads messages from the BlockingCollection, parses them, and forwards them to an IoT Hub. But again, the receiving messages were interrupted.**

We used Windows device (Azure Windows VM - Windows 10 Pro OS), Windows container, and IoT Edge 1.0.10 for Windows but the IoT Edge module stopped receiving data. After that, we upgraded the IoT Edge to 1.1 (https://docs.microsoft.com/en-us/azure/iot-edge/how-to-install-iot-edge-windows-on-windows?view=iotedge-2018-06) but the same thing happens.

**But today I saw that the interruption occurs at the same time, 2:01 AM or 02:02 AM!**

I checked the IoT Hub diagram for the last twenty days and attached screenshots for the last four days. **After the interruption, I usually redeployed the IoT Edge module on the device and the module received messages again. Rarely it happens that receiving messages starts slowly again and redeployment is not needed.**

Why does the interruption always occur at the same time? It's as if some Azure planned maintenance operation interrupts receiving messages by the module? The interruption is not related to the number of messages, because even on weekends there is an interruption at the same time when there are significantly fewer messages.

### Output of `iotedge check`
We use a temporary certificate but we will create another certificate for the production.
![IoT Edge Check](https://user-images.githubusercontent.com/28098446/120275886-4014bc00-c2b2-11eb-89a3-540dac9eecfd.png)

### Device Information
* Host OS [Windows 10 Pro]: 
* Container OS [Windows containers]: 

### Runtime Versions
• iotedge: 1.1.1
• Edge Agent: 1.1.2.41361617
• Edge Hub: 1.1.2.41361617

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs
Logs for 05-23-2021:

[support_bundle_1_23.05.2021.zip](https://github.com/Azure/iotedge/files/6573909/support_bundle_1_23.05.2021.zip)

<summary>Screenshot of IoT digram for 05-23-2021</summary>

![2021-05-23](https://user-images.githubusercontent.com/28098446/120278662-d8607000-c2b5-11eb-96c6-e4f809297e80.png)

## Additional Information
Other screenshots are attached:
![IoT Hub - 2021-05-28](https://user-images.githubusercontent.com/28098446/120278829-0cd42c00-c2b6-11eb-91d7-64ba7c881bd0.png)
![IoT Hub - 2021-05-29](https://user-images.githubusercontent.com/28098446/120278840-0fcf1c80-c2b6-11eb-85d8-afceb7651cf9.png)
![IoT Hub - 2021-05-30](https://user-images.githubusercontent.com/28098446/120278841-1067b300-c2b6-11eb-9f79-2501132dae4f.png)
![IoT Hub - 2021-05-27](https://user-images.githubusercontent.com/28098446/120278843-11004980-c2b6-11eb-91b1-0a6f26fa5d08.png)

**Comment by draganrmijailovic** [comment]
But there is another "Device to cloud message" diagram and **currently, there are no new messages on it**. Screenshots are added. I checked the IoT Edge module, there are not newly received messages. **My mistake, I have seen interruptions in different periods based on the "Device to cloud message" diagram.**
![IoT Edge - Device to cloud messages - Interruption - 2021-05-29 - 2021-05-30](https://user-images.githubusercontent.com/28098446/120370391-d37ed900-c314-11eb-8338-54ce5d4662e5.png)
![IoT Edge - Device to cloud messages - Interruption - 2021-06-01](https://user-images.githubusercontent.com/28098446/120370399-d5489c80-c314-11eb-89c6-e1875f56c4ed.png)

**I checked the IoT Edge log module and currently, there are newly
[support_bundle_2021-06-01_Last Day.zip](https://github.com/Azure/iotedge/files/6579763/support_bundle_2021-06-01_Last.Day.zip)
 received messages.**

The logs in the last 24 hours:

**Comment by draganrmijailovic** [comment]
We improved our C# program for receiving messages via UDP protocol in the previous period. I briefly described it in the initial comment. We consulted with the Pacific Track team. They have a similar program in Java and they told us that our program is corrected. 
But, is our Azure infrastructure OK? We use an IoT Edge Windows device (Azure VM with Windows Pro 10 OS and IoT Edge 1.1). Should we migrate our IoT Edge device to Linux? We created the device based on the https://docs.microsoft.com/en-us/azure/iot-edge/how-to-install-iot-edge-windows-on-windows?view=iotedge-2018-06 but maybe this solution should not be used in production and should we use another solution?

**Comment by yophilav** [comment]
Hello! 
Thank you for the updates! I'm taking a look right now.

**Comment by draganrmijailovic** [comment]
Hello,
thanks a lot. The following is interesting: we have been receiving data for several months without interruption via a basic version of the program (with one thread). The interruption only occurred at the beginning when the Windows update was performed unplanned (after we turned off Windows update) and when the temporary certificate was expired. 
Now interruption occurs after a day (in the last month). The number of devices is not increased and the number of received messages is the same as before. We improved the program but without effect.
I hope that we will get some information from you and IoT Edge runtime logs.

**Comment by yophilav** [comment]
I found a few funny behavior going on with the edgeAgent here.

First, it seems like somehow the custom image was not properly publish to container registry? EdgeAgent is having a hard time trying to deploy the `tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64` image.
```
<7> 2021-06-01 07:49:31.094 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.Core.PlanRunners.OrderedRetryPlanRunner] - Skipping command "Command Group: (
  [Command Group: (
  [Prepare update module IotEdgeModuleReceivingDataFromPt40]
  [Stop module IotEdgeModuleReceivingDataFromPt40]
  [Update module IotEdgeModuleReceivingDataFromPt40]
)]
  [Start module IotEdgeModuleReceivingDataFromPt40]
)" in deployment 153. Error: Error calling prepare update for module IotEdgeModuleReceivingDataFromPt40: Could not prepare update for module "IotEdgeModuleReceivingDataFromPt40"
	caused by: Could not pull image tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64
	caused by: manifest for tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64 not found: manifest unknown: manifest tagged by "0.1.18-windows-amd64" is not found. Will retry in 00m:14s

```
For this issue, Could you please verify that you can correctly pull an image on the device using
`docker pull tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64` ?

-------------------------------------------------------------------------------

Second, edgeAgent is having a hard time trying to communicate with the underlying management socket (API call to edgelet). It is weird that the edgeAgent claims to have a hard time communicating to the edgelet, it proceeded to list the "current" running module which includes the module `iotedgemodulereceivingdatafrompt40:0.1.19-windows-amd64`. I'm confused as how it gets into this state. 

- What is the tag of `tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40` that you want to be deployed?
- How did you deploy the `iotedgemodulereceivingdatafrompt40:0.1.19-windows-amd64` ?

I wonder that because the edgeAgent is repeatedly trying to "redeploy" an existing module name with a different tag may have some effects as an interruption.... (since the log suggested that the stop-command was triggered against the module).

```
<7> 2021-05-31 18:08:03.780 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient] - Error when getting an Http response from unix:///C:/ProgramData/iotedge/mgmt/sock for List identities
System.TimeoutException: Operation timed out
   at Microsoft.Azure.Devices.Edge.Util.TaskEx.TimeoutAfter[T](Task`1 task, TimeSpan timeout) in C:\agent\_work\2\s\edge-util\src\Microsoft.Azure.Devices.Edge.Util\TaskEx.cs:line 128
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in C:\agent\_work\2\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\versioning\ModuleManagementHttpClientVersioned.cs:line 146
<7> 2021-05-31 18:08:03.782 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleIdentityLifecycleManager] - Error getting module identities.
System.TimeoutException: Operation timed out
   at Microsoft.Azure.Devices.Edge.Util.TaskEx.TimeoutAfter[T](Task`1 task, TimeSpan timeout) in C:\agent\_work\2\s\edge-util\src\Microsoft.Azure.Devices.Edge.Util\TaskEx.cs:line 128
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in C:\agent\_work\2\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\versioning\ModuleManagementHttpClientVersioned.cs:line 146
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in C:\agent\_work\2\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\versioning\ModuleManagementHttpClientVersioned.cs:line 146
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2020_07_07.ModuleManagementHttpClient.GetIdentities() in C:\agent\_work\2\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\version_2020_07_07\ModuleManagementHttpClient.cs:line 93
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleIdentityLifecycleManager.GetModuleIdentitiesAsync(Diff diff) in C:\agent\_work\2\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\ModuleIdentityLifecycleManager.cs:line 90
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleIdentityLifecycleManager.GetModuleIdentitiesAsync(ModuleSet desired, ModuleSet current) in C:\agent\_work\2\s\edge-agent\src\Microsoft.Azure.Devices.Edge.Agent.Edgelet\ModuleIdentityLifecycleManager.cs:line 47
   <7> 2021-05-31 18:08:03.783 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.Core.Planners.HealthRestartPlanner] - List of desired modules is - {"edgeHub":{"status":"running","restartPolicy":"always","version":"","imagePullPolicy":"on-create","type":"docker","settings":{"image":"mcr.microsoft.com/azureiotedge-hub:1.1","createOptions":"{\"HostConfig\":{\"PortBindings\":{\"5671/tcp\":[{\"HostPort\":\"5671\"}],\"8883/tcp\":[{\"HostPort\":\"8883\"}],\"443/tcp\":[{\"HostPort\":\"443\"}]}}"},"env":{}},"edgeAgent":{"version":"","status":"running","restartPolicy":"always","imagePullPolicy":"on-create","startupOrder":0,"type":"docker","settings":{"image":"mcr.microsoft.com/azureiotedge-agent:1.1","createOptions":"{}"},"env":{}},"IotEdgeModuleReceivingDataFromPt40":{"status":"running","restartPolicy":"always","version":"1.0.0","imagePullPolicy":"on-create","type":"docker","settings":{"image":"tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64","createOptions":"{\"ExposedPorts\":{\"1208/udp\":{}},\"HostConfig\":{\"PortBindings\":{\"1208/udp\":[{\"HostPort\":\"1208\"}]}}"},"env":{}}
<7> 2021-05-31 18:08:03.784 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Agent.Core.Planners.HealthRestartPlanner] - List of current modules is - {"edgeHub":{"restartPolicy":"always","exitCode":0,"statusDescription":"running","lastStartTimeUtc":"2021-05-29T05:50:10.2460139","lastExitTimeUtc":"2021-05-29T05:49:15.9686542","restartCount":0,"lastRestartTimeUtc":"2021-05-29T05:49:15.9686542","runtimeStatus":"running","status":"running","imagePullPolicy":"on-create","type":"docker","settings":{"image":"mcr.microsoft.com/azureiotedge-hub:1.1","imageHash":"sha256:3d69e472477c4baa84782323ed074cd51a856ff1e5b75f1a5f59c024db0205c3","createOptions":"{\"HostConfig\":{\"PortBindings\":{\"5671/tcp\":[{\"HostPort\":\"5671\"}],\"8883/tcp\":[{\"HostPort\":\"8883\"}],\"443/tcp\":[{\"HostPort\":\"443\"}]}}"},"env":{}},"edgeAgent":{"exitCode":0,"statusDescription":"running","lastStartTimeUtc":"2021-05-29T05:50:02.7006807","lastExitTimeUtc":"2021-05-29T05:49:15.597914","runtimeStatus":"running","imagePullPolicy":"on-create","startupOrder":0,"type":"docker","settings":{"image":"mcr.microsoft.com/azureiotedge-agent:1.1","imageHash":"sha256:476e5f37b3664b0430e1b1590c0422d11e737a88f9dc399108aa84c58aaf1eaa","createOptions":"{\"Labels\":{\"net.azure-devices.edge.create-options\":\"{}\",\"net.azure-devices.edge.env\":\"{}\",\"net.azure-devices.edge.owner\":\"Microsoft.Azure.Devices.Edge.Agent\"}"},"env":{}},"IotEdgeModuleReceivingDataFromPt40":{"exitCode":0,"statusDescription":"running","lastStartTimeUtc":"2021-05-30T09:36:05.8286917","lastExitTimeUtc":"0001-01-01T00:00:00","restartCount":0,"lastRestartTimeUtc":"0001-01-01T00:00:00","runtimeStatus":"running","version":"1.0.0","status":"running","restartPolicy":"always","imagePullPolicy":"on-create","type":"docker","settings":{"image":"tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.19-windows-amd64","imageHash":"sha256:581b1be7c6bf127f6933eb766676178bbbb4bb0e41037f1c73d2a246550cba74","createOptions":"{\"ExposedPorts\":{\"1208/udp\":{}},\"HostConfig\":{\"PortBindings\":{\"1208/udp\":[{\"HostPort\":\"1208\"}]}}"},"env":{}}
```

**Comment by draganrmijailovic** [comment]
Thanks for your answer.

- _Could you please verify that you can correctly pull an image on the device using docker pull tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64?
(If that fails, could you upload the image and see if the issue is resolved in edgeAgent logs please?)_
This command fails (the screenshot is attached). It is the old version of the IoT Edge module that should not be used. Should I create and deploy it on the device? But, the device should use the 0.1.19 version of the image.
The error message after performing the pull command: "The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running."

- _What is the tag of tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40 that you want to be deployed?_

The tag is 0.1.19-windows-amd64 (the screenshots are attached)

- _How did you deploy the iotedgemodulereceivingdatafrompt40:0.1.19-windows-amd64?_

I deployed the iotedgemodulereceivingdatafrompt40:0.1.19-windows-amd64 via Visual Studio. But, our team started deploying the IoT edge module via Terraform (I think that when the code or the version is changed it deploys the IoT module) as well. But, when I was resolving the problem I was publishing the image via Visual Studio. I previously deleted the old version and pushed the new one via Visual Studio. Probably there were situations when the new version of the image is deployed without removing the old one (I think that Terraform does not delete the old version).

But, should I recreate the device due to the old versions of the module? Temporary we can disable deploying via Terraform as well.
The screenshots:
![Pull command - Error](https://user-images.githubusercontent.com/28098446/121393015-36452580-c950-11eb-923d-40d04e72bb39.PNG)
![iotedge list](https://user-images.githubusercontent.com/28098446/121393029-38a77f80-c950-11eb-971e-80e629c451e7.PNG)
<img width="707" alt="iotedgemodulereceivingdatafrompt40 Repository - 0 1 19-windows-amd64 Tag" src="https://user-images.githubusercontent.com/28098446/121393044-3b09d980-c950-11eb-85ab-6b0ae8c71803.PNG">

**Comment by yophilav** [comment]
@draganrmijailovic

1. For the docker command: 
`docker -H npipe:////./pipe/iotedge_moby_engine tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.18-windows-amd64`
This will probably fail since you don't have the image available on the Container Registry. 

2. What is `Terraform` and how is it used in relation to iotedge? I suspect the fact that edgeAgent is repeated trying to redeploy the `iotedgemodulereceivingdatafrompt40` is why the module keep getting interrupted. Now the question is... why is edgeAgent is still trying to deploy the invalid `0.1.18-windows-amd64` even though the `0.1.19-windows-amd64` is deployed -- does someone triggering it to do so?

**Comment by draganrmijailovic** [comment]
You are right. When I go to Set Modules via the Azure portal I see the 0.1.18 version. The screenshot is attached. But via the IoT edge list command I get 0.1.19.
<img width="837" alt="Set Modules - iotedgemodulereceivingdatafrompt40 Repository - 0 1 18 - windows-amd64 Tag" src="https://user-images.githubusercontent.com/28098446/121396936-27607200-c954-11eb-8664-05c5e3ebd4ad.PNG">

**Comment by draganrmijailovic** [comment]
Our team wants to automatize managing Azure resources via Terraform. That's the only reason. And they included the IoT resources in Terraform: https://learn.hashicorp.com/collections/terraform/azure-get-started.

**Comment by yophilav** [comment]
@draganrmijailovic 

Thank you for the response! 
This is an odd behavior because when you do the deployment via VSCode, it _should_ automagically update the portal to be using the latest `0.1.19` version.

1. Could you please give me a detail steps on how exactly you deployed the `0.1.18` and `0.1.19` (what tools/commands you were using?, etc) This possibly may be a bug on either the VSCode tool or portal issue.
2. As a workaround, you could just update the portal to be version `0.1.19`. Then we can observe the edgeAgent log if the error goes away.

**Comment by draganrmijailovic** [comment]
1. I don't use VS Code and I use Visual Studio: I change the version in module.json and execute the Build And Push IoT Edge Modules command (the right click on the project). After pushing the IoT module in Azure registry I choose to Create Deployment. The screenshots:
<img width="905" alt="Visual Studio - Build and Push" src="https://user-images.githubusercontent.com/28098446/121404077-e1a7a780-c95b-11eb-9259-7b311b586367.png">
<img width="905" alt="Visual Studio - Create Deployment" src="https://user-images.githubusercontent.com/28098446/121404089-e40a0180-c95b-11eb-8e0a-59db7f632842.png">
2. I updated the version from the portal but the IoT module did not start receiving messages - there are no new logs. **After that, I deleted the IoT module and repeated the described procedure Build and Push and Create deployment but 0.1.18 IoT module is deployed again! The module is in a failed state. **
<img width="905" alt="Container registry and IoT Edge module" src="https://user-images.githubusercontent.com/28098446/121405149-fafd2380-c95c-11eb-8d70-cd7976aa1c1e.png">

**Comment by yophilav** [comment]
@draganrmijailovic 
It looks like edgeAgent is still receiving the deployment for `0.1.18` from the IotHub.  
How do you deploy the module from the portal? 
Could you follow the 1-5 steps from this doc
https://docs.microsoft.com/en-us/azure/iot-edge/how-to-deploy-modules-portal?view=iotedge-2020-11#select-device-and-add-modules
then for step 6 onwards do the follow
(6) Click "IotEdgeModuleReceivingDataFromPt40" entry
(7) Modify "Image URI" to "tmsstagingacr.azurecr.io/iotedgemodulereceivingdatafrompt40:0.1.19-windows-amd64"
(8) Wait for about a minute & observe edgeAgent if the error is still showing up.

**Comment by draganrmijailovic** [comment]
I followed the steps and updated the IoT module to 0.1.19. **And the module is receiving messages again!** But I'll see if there are any interruptions in the next period.

**Comment by yophilav** [comment]
Awesome! Would you mind observe the edgeAgent log one more time please? This should ensure that it is working properly... As said, you should no longer see the `pull` error. 

This is possibly an issue with an extension on Visual Studio add on that deploys the module (which somehow didn't update the IoTHub to the proper version). Would you mind trying to basically deploy another say... `0.1.20` (using the same code base) and see if the IoTHub is updated to the proper version please?

**Comment by draganrmijailovic** [comment]
Hm... **Version is returned to 0 1 18! And pull error happened in the last one hour.** I attached the logs and screenshot.
<img width="767" alt="Version is returned to 0 1 18" src="https://user-images.githubusercontent.com/28098446/121426379-d06a9500-c973-11eb-8107-a4a08796028e.PNG">
[support_bundle_2021-06-09 19 26 11.zip](https://github.com/Azure/iotedge/files/6626790/support_bundle_2021-06-09.19.26.11.zip)

I will repeat the earlier steps for updating the version.

**Comment by draganrmijailovic** [comment]
Now I updated the version to 0.1.19 again. The screenshot is attached. How was the version reverted to 0.1.18 ...
<img width="939" alt="Version is returned to 0 1 19" src="https://user-images.githubusercontent.com/28098446/121427213-bbdacc80-c974-11eb-8e16-9f88bfff3959.PNG">

**Comment by draganrmijailovic** [comment]
**But I can already see that the desired value 0.1.18 and the reported value is 0.1.19!** The screenshot is attached. 
<img width="934" alt="Desired and Reported Value" src="https://user-images.githubusercontent.com/28098446/121428213-d792a280-c975-11eb-8f7d-343fea715838.PNG">

**Comment by yophilav** [comment]
> But I can already see that the desired value is 0.1.18 and the reported value is 0.1.19! The screenshot is attached.

Urhg. I'm not sure how to resolve this. You should check on who is triggering a deployment to 0.1.18 version. The reported value of 0.1.19 is simply saying that 0.1.19 is running.. but it doesn't mean it won't get interrupted by repeated 0.1.18 deployment.

Ps. Are you using layer deployment somewhere? Somehow someone is triggering a deployment... Until you find THE someone, it's likely that you will run into an interruption.

**Comment by draganrmijailovic** [comment]
Now the version is already changed to 0.1.18 and yes, probably receiving messages will be interrupted again in the next period ...

I know for Azure pipeline that uses Terraform for automatic redeploying IoT Edge module when its C# project has some changes.  The pipeline is performed only after approving a pull request. I don't know the details but I saw that this pipeline was performed on May 28. It has a few tasks for pushing the image to the registry and deploying it on the IoT Edge device. But it certainly didn't perform a while ago or after manually redeploying the module from Visual Studio. It triggers only after changes in the module's C# project, creating and approving PR. Probably the problem is the IoT Edge device and its runtime configuration. But, I don't know how its behavior can be influenced.

**Comment by yophilav** [comment]
Maybe you can try clicking around IoTHub to see if there is a lingering deployment queue up somewhere?  I don't really what else I can do. It looks like the root cause comes from the repeated module redeployment.

**Comment by draganrmijailovic** [comment]
I found a deployment that references the 0.1.18 version on the portal and removed it. After that, the version of the module is not reverted to 0.1.18 and the version remained 0.1.19 :) 
I'll see if that solves the problem of interruption and I'll check logs. 
Thanks for your cooperation.

**Comment by draganrmijailovic** [comment]
Receiving messages was interrupted after a few hours. But, there are not Pull errors in log files. I attached them. 
1. Do you see anything else interesting in the logs?
[support_bundle_since_2021-06-10 11 09 33 UTC.zip](https://github.com/Azure/iotedge/files/6633284/support_bundle_since_2021-06-10.11.09.33.UTC.zip)

2. I will revert the initial simplified version of the module (with one thread) that worked a few months without interruptions.

**Comment by veyalla** [comment]
@draganrmijailovic We just released an [IoT Edge monitoring solution](https://aka.ms/iotshow/edgemon) in public preview. It lets you visualize edgeHub metrics that might provide some clues on what is going on. See https://aka.ms/edgemon-docs

**Comment by draganrmijailovic** [comment]
Ok, thanks for the links and information!

**Comment by yophilav** [comment]
@draganrmijailovic 
I saw one transient hiccup at the beginning but it recovered itself immediately. The rest of communication seems to be quite clean.

**Comment by draganrmijailovic** [comment]
Ok, thanks for the answer. I will see whether there will be interruptions with the old version of the program for receiving data via UDP protocol.

**Comment by yophilav** [comment]
@draganrmijailovic Any update on the issue? Is it alright if I close the issue?

**Comment by draganrmijailovic** [comment]
@yophilav You can close the issue. I implemented another C# program for receiving messages via UDP protocol and for now there weren't any interruptions. I will apply the IoT Edge monitoring solution in the next period.  That will be very useful. 
Thanks for your cooperation!

**Comment by yophilav** [comment]
@draganrmijailovic  
Sweet! Thank you for all your help!
You have a nice day :)

**Comment by draganrmijailovic** [comment]
@yophilav Unfortunately, after almost two days there was an interruption again! But, you checked your side and there is no issue. I will try to investigate the problem by IoT Edge monitoring solution. In the end, if I fail, I will have to apply other solutions for hosting the program for receiving messages via UDP protocol.   
Probably Azure Windows Pro 10 VM is not a good choice for the IoT Edge device, I am not sure.

**Comment by draganrmijailovic** [comment]
@yophilav Can you open the issue again? And I will try to set up an IoT Edge monitoring solution. Probably then we will come to additional information about the issue.

**Comment by yophilav** [comment]
> you checked your side and there is no issue. I will try to investigate the problem by IoT Edge monitoring solution. In the end,

Could you provide us with the support-bundle? (Note: the bundle may contain secrets variable as a print out which you may want to reduct)

**Comment by draganrmijailovic** [comment]
1. I failed to return the logs by the support-bundle command. This command hanged. But, all modules were in Running status. The screenshots are attached. I could return logs by the following command: iotedge logs edgeAgent --since 1h and I can show the logs.
2. @veyalla I installed an IoT Edge monitoring solution. It looks great. I hope we can come up with some additional information using it. The screenshots
![Iot Edge support-bundle command hangs](https://user-images.githubusercontent.com/28098446/122588869-e48f5000-d05f-11eb-8759-f219da981303.PNG)
![iotedge list - All modules in running status](https://user-images.githubusercontent.com/28098446/122588882-e6f1aa00-d05f-11eb-9e35-ae7a2e48434a.PNG)
![IoT Edge Fleet View](https://user-images.githubusercontent.com/28098446/122588890-e9ec9a80-d05f-11eb-9dd9-bea2c7df7ad2.PNG)
![IoT Edge Device Details](https://user-images.githubusercontent.com/28098446/122588902-ece78b00-d05f-11eb-8375-922872fab2cd.PNG)
 are added.

**Comment by veyalla** [comment]
One thing that stands out from the metrics is the disk usage is close to 100%. Perhaps the device runs out of disk space after running for a while? 

![gh-issue](https://user-images.githubusercontent.com/22105760/122592393-0f09ea80-d019-11eb-983a-14f8795d6658.png)

**Comment by draganrmijailovic** [comment]
@veyalla Thanks for your answer.

I checked disk space and there is free space: 97.5 GB. The screenshot is attached. But, probably logs are written on a disk or other activities are performed, and its usage is high...? The IoT Edge module only forwards messages to the cloud after parsing them to JSON in the memory.

But, after 12 hours the interruption happened again. The device was unhealthy. I saw high CPU usage. The screenshot is added. I increased disk and CPU performance in VM. 

![Disk Space](https://user-images.githubusercontent.com/28098446/122613113-93458780-d084-11eb-9038-86c3e22c7d50.PNG)
![Device - Unhealthy](https://user-images.githubusercontent.com/28098446/122613120-950f4b00-d084-11eb-9266-8e9a04c3e704.PNG)

**Comment by veyalla** [comment]
Interesting, not sure why metrics are reporting high disk usage if there is this much space available..you can get more information from [host tab ](https://docs.microsoft.com/azure/iot-edge/how-to-explore-curated-visualizations?view=iotedge-2020-11&tabs=devices%2Chost#iot-edge-device-details-workbook) in device's detailed metrics.

For high CPU utilization, you can check [modules](https://docs.microsoft.com/azure/iot-edge/how-to-explore-curated-visualizations?view=iotedge-2020-11&tabs=devices%2Cmodules#iot-edge-device-details-workbook) tab to see if there is increased usage from any specific module and see the [live logs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-explore-curated-visualizations?view=iotedge-2020-11&tabs=devices%2Clivelogs#iot-edge-device-details-workbook) (especially from edgeHub)  around the time the CPU usage ramped up (click and drag on the chart to zoom into the time range)..

**Comment by draganrmijailovic** [comment]
And host tab shows free disk space:

![Host tab](https://user-images.githubusercontent.com/28098446/122619540-f5a48500-d090-11eb-964e-f3ba5375fed1.PNG)

**But, disk usage is still high!**

![PT-40 - Flat View](https://user-images.githubusercontent.com/28098446/122620522-78c6da80-d093-11eb-9138-06f35da689e6.PNG)


Modules tab - Resource consumption - Our module loads the CPU. But, after improving CPU its usage is reduced but probably still high ... But message processing is demanding, there is a parser that processes messages' byte arrays ... Otherwise, processing threads sleep for a time if there are no new messages for processing. The main thread is blocked and the infinity loop is not used. 

![Modules tab - Resource consumption](https://user-images.githubusercontent.com/28098446/122619559-fe955680-d090-11eb-8d7b-2c0165dfd581.PNG)

Live tab - I chose time intervals with high CPU usage and edgeHub minimally loads the CPU. I think that our module only loads CPU. 

I will see whether there will be interruptions when CPU and disk performance are increased. I'll let you know the result.

**Comment by draganrmijailovic** [comment]
@veyalla
1. There was no interruption after one day and increasing CPU and disc performance. Now CPU usage is 75% constantly regardless of the number of received messages:
![Messages vs CPU usage](https://user-images.githubusercontent.com/28098446/122647418-719bdd00-d124-11eb-8be7-e90959c55400.PNG)
When I go to VM's Task Manager CPU usage is 80% - 33% (Task Manager) = 47%. But there are other processes that load CPU and they are not related to IoT Edge runtime (the second screenshot)  
2. Disc usage on IoT Edge Flat View is 100% constantly but when I go to Task Manager of VM disc usage is 0% or 0.1%:
![VM - Task Manager (CPU, Disc,  )](https://user-images.githubusercontent.com/28098446/122647519-d9eabe80-d124-11eb-9c83-09620a0a4af8.PNG)
Is there an inconsistency?

**Comment by draganrmijailovic** [comment]
@veyalla
The interruption happened again. CPU usage was increased to 100% before it. I propose the problem in our module/UDP listener for receiving messages. We tried several versions of the module/UDP listener. The current version of the UDP listener is implemented by one company that has a large number of clients. But, it doesn’t work well for us ...   

![Interruption - Device health](https://user-images.githubusercontent.com/28098446/122664096-84072c80-d19f-11eb-8eb7-3ba450432232.PNG)
![Interruption - Messages to cloud](https://user-images.githubusercontent.com/28098446/122664100-88334a00-d19f-11eb-9229-2f5d7cbaafd4.PNG)
![Interruption - CPU usage](https://user-images.githubusercontent.com/28098446/122664105-8a95a400-d19f-11eb-8a77-0b6c30a2632a.PNG)
![Interruption - Messaging](https://user-images.githubusercontent.com/28098446/122664109-8c5f6780-d19f-11eb-80ab-02a25dbfe15f.PNG)

**Comment by draganrmijailovic** [comment]
We have increased the performance of the processor and for now, there were no interruptions. CPU usage is 25% regardless of the number of received messages. But, CPU has 8 cores and monthly costs will be high. We don't sure whether this solution will be scalable for the larger number of devices.
1. We will investigate our listener and we will try to improve it.
2. @veyalla Will CPU usage decrease if we switch to Linux instead of Windows Pro 10 OS and IoT Edge 1.2 instead of IoT Edge 1.1? Should we try that?

**Comment by veyalla** [comment]
Since the CPU spike is coming from a custom module, I'm not sure if Linux containers will help or not. However, given the [roadmap for Windows containers](https://docs.microsoft.com/en-us/azure/iot-edge/support?view=iotedge-2018-06#windows-containers) I would not recommend Windows containers for any new projects. 

Regardless, it is always recommended to set[ CPU/memory limits on modules](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-use-create-options?view=iotedge-2020-11#restrict-module-memory-and-cpu-usage) to prevent these kind of situations.

**Comment by draganrmijailovic** [comment]
OK, thanks for the information. I will set CPU/memory limits on modules.

**Comment by yophilav** [comment]
@draganrmijailovic 
How is the issue going? May I close the issue now?

**Comment by yophilav** [comment]
I'm going to close the issue. Please feel free to reopen as necessary. Thank you :)

**Comment by draganrmijailovic** [comment]
@yophilav OK. I created the new Linux IoT Edge device with IoT Edge 1.2 runtime. But, everything was the same, the interruption was after about twenty hours. After I excluded the parser from the UDP listener that we got from a manufacturer and the interruption happened after six days. And at the end, I changed the UDP listener and excluded the parser, and CPU usage has been 0.1% for four days and no interruptions.
